[thisvid] Add extractor (closes #17713)

jhwgh1968 opened the pull request:
### Before submitting a *pull request* make sure you have:
- [x] At least skimmed through [adding new extractor tutorial](https://github.com/ytdl-org/youtube-dl#adding-support-for-a-new-site) and [youtube-dl coding conventions](https://github.com/ytdl-org/youtube-dl#youtube-dl-coding-conventions) sections
- [x] [Searched](https://github.com/ytdl-org/youtube-dl/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests
- [x] Checked the code with [flake8](https://pypi.python.org/pypi/flake8)

### In order to be accepted and merged into youtube-dl each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check one of the following options:
- [x] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- [ ] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)

### What is the purpose of your *pull request*?
- [ ] Bug fix
- [ ] Improvement
- [x] New extractor
- [ ] New feature

---

### Description of your *pull request* and other information

Adds a new extractor for adult website ThisVid.

A previous attempt was made long ago in #16009, but required a new library to execute javascript on the page. This version uses PhantomJS, which is already used by another extractor.

Closes #17713.

dstftw commented on youtube_dl/extractor/thisvid.py:
```diff
@@ -0,0 +1,109 @@
+        # Embedded versions contain no metadata, so simply fetch their main
+        # page instead
+        if '/embed/' in url:
+            return self.extract(
```
No, add another extractor and properly delegate.

dstftw commented on youtube_dl/extractor/thisvid.py:
```diff
@@ -0,0 +1,109 @@
+    AUTHOR_HREF_RE = re.compile(r'<a.*class="author"[^>]*>([^<]+)</a')
+    TITLE_RE = re.compile(r'<title>(?P<title>.*) - ThisVid.com</title>')
+    HREF_RE = re.compile(r'href="([^"]+)"')
+    LINK_RE = re.compile(r'<link rel="canonical" href="(?P<href>[^"]+)"')
```
Inline everything used only once.

dstftw commented on youtube_dl/extractor/thisvid.py:
```diff
@@ -0,0 +1,109 @@
+    _TESTS = [
+        {
+            'url': 'https://thisvid.com/videos/' +
+                   'madonna-show-in-sexy-underwear/',
```
Don't carry URLs. Read coding conventions.

dstftw commented on youtube_dl/extractor/thisvid.py:
```diff
@@ -0,0 +1,109 @@
+
+
+class ThisVidIE(InfoExtractor):
+    _VALID_URL = r'https?://(?:www\.)?thisvid\.com/(videos|embed)/' + \
```
Don't capture groups you don't use.

jhwgh1968 commented:
I believe I have fixed everything, @dstftw.

jhwgh1968 commented:
Ping @dstftw, is there anything else I need to fix? Or can this be merged?

ThisVidLover commented:
@jhwgh1968 
I think You should just go ahead & merge this Yourself!
If @dstftw still has something to criticize, I'm sure he will show up again!

jhwgh1968 commented:
I'm not sure what you mean, @ThisVidLover. Only @dstftw, @remitamine, and @yan12125 can merge things as far as I know, and @remitamine does not seem to do many PRs.

I'm already using it internally. This is my attempt to give back to the open source community.

But since this PR has sat without review for almost a month, I may give up... :disappointed:

jhwgh1968 commented:
Ping @dstftw or @remitamine, I just rebased this and added another case that was missing. Can I get a code review?

dstftw commented:
I've already pointed out: you must properly delegate from one extractor to another.

jhwgh1968 commented:
Apparently I didn't understand what you meant. Thanks for getting back to me.

Hopefully this version should fix that.

dstftw commented on youtube_dl/extractor/thisvid.py:
```diff
@@ -0,0 +1,111 @@
+            });
+            console.log('---'); // ensure any errors appear above where we are
+            console.log(fixed_url);
+            console.log(thumbnail_url);
```
Just output complete stringified flashvars and consume in python code as JSON.

jhwgh1968 replied:
I now output the flashvars json, but still need to get rid of the error output that spews above it.

dstftw commented on youtube_dl/extractor/thisvid.py:
```diff
@@ -0,0 +1,111 @@
+        phantom = PhantomJSwrapper(self, required_version='2.0')
+        webpage, output = phantom.get(url, html=webpage, jscode=jscode)
+        video_url, thumbnail_url = \
+            output.split("---", 2)[1].strip().split("\n", 2)
```
Parse from flashvars JSON.

dstftw commented on youtube_dl/extractor/thisvid.py:
```diff
@@ -0,0 +1,111 @@
+        video_url, thumbnail_url = \
+            output.split("---", 2)[1].strip().split("\n", 2)
+        assert not video_url.startswith('function'), "old flashvars found"
+        video_url = video_url.split("?", 2)[0].strip("/")
```
What's the point of this? Remove.

jhwgh1968 replied:
The URL sometimes has extra stuff at the end. I can refactor this as part of the change above.

jhwgh1968 replied:
Added comment

dstftw commented on youtube_dl/extractor/thisvid.py:
```diff
@@ -0,0 +1,111 @@
+        video_url = video_url.split("?", 2)[0].strip("/")
+        if thumbnail_url.startswith("//"):
+            thumbnail_url = 'https:' + thumbnail_url
+        video_id = video_url.split("/")[-2]
```
Parse from flashvars JSON.

dstftw commented on youtube_dl/extractor/thisvid.py:
```diff
@@ -0,0 +1,111 @@
+
+        # parse the title information
+        title = self._search_regex(
+            r'<title>(?P<title>.*) -([a-zA-Z ]+ at)? ThisVid(\.com| tube)</title>',
```
Don't capture empty title.

dstftw commented on youtube_dl/extractor/thisvid.py:
```diff
@@ -0,0 +1,111 @@
+
+        # parse the author information from a profile link
+        m = re.search(
+                r'<a.*class="author" href="(?P<url>[^"]+)"[^>]*>(?P<id>[^<]+)</a',
```
Fix indent.

jhwgh1968 replied:
PEP-8 did not flag this. Do you want it aligned with the parenthesis then?

jhwgh1968 replied:
Refactored out of existence.

dstftw commented on youtube_dl/extractor/thisvid.py:
```diff
@@ -0,0 +1,111 @@
+        # parse the author information from a profile link
+        m = re.search(
+                r'<a.*class="author" href="(?P<url>[^"]+)"[^>]*>(?P<id>[^<]+)</a',
+                webpage)
```
Breaks extraction if not available. Again read coding conventions.

jhwgh1968 replied:
I can fix the assert.

I could not see how to return more than one group from ``_search_regex``. Is there a way to do that?

jhwgh1968 replied:
I put in two separate ``_search_regex`` calls. It would be nice if I could do it with one.

jhwgh1968 commented:
There were a couple things that were not clear in the comments. For those things, I added more comment blocks or tried refactoring it to be better.

Hopefully I've addressed everything, @dstftw.

jhwgh1968 commented:
Ping @dstftw, is there anything else I need to fix?

jhwgh1968 commented:
Ping @dstftw, I think I did everything you wanted. Did I miss anything?

rigstot commented:
Hi guys (@jhwgh1968, @dstftw & co) -- I've had a shot of a thisvid extractor myself and managed to get rid of the phantomjs dependency. -> #21651

I'm fairly certain that mine will be possible to adapt for any KVS player / kernel-scripts.com based video site, as I've reverse engineered their license validator.

jhwgh1968 commented:
Good to know, @rigstot! I'll close this, and comment on yours.
